File: python/semantic_kernel/functions/kernel_function_from_method.py
# Copyright (c) Microsoft. All rights reserved.

import inspect
import logging
from collections.abc import Callable
from inspect import isasyncgen, isasyncgenfunction, isawaitable, iscoroutinefunction, isgenerator, isgeneratorfunction
from typing import Any

from pydantic import Field, ValidationError

from semantic_kernel.exceptions import FunctionExecutionException, FunctionInitializationError
from semantic_kernel.filters.functions.function_invocation_context import FunctionInvocationContext
from semantic_kernel.functions.function_result import FunctionResult
from semantic_kernel.functions.kernel_function import KernelFunction
from semantic_kernel.functions.kernel_function_metadata import KernelFunctionMetadata
from semantic_kernel.functions.kernel_parameter_metadata import KernelParameterMetadata

logger: logging.Logger = logging.getLogger(__name__)


class KernelFunctionFromMethod(KernelFunction):
    """Semantic Kernel Function from a method."""

    method: Callable[..., Any] = Field(exclude=True)
    stream_method: Callable[..., Any] | None = Field(default=None, exclude=True)

    def __init__(
        self,
        method: Callable[..., Any],
        plugin_name: str | None = None,
        stream_method: Callable[..., Any] | None = None,
        parameters: list[KernelParameterMetadata] | None = None,
        return_parameter: KernelParameterMetadata | None = None,
        additional_metadata: dict[str, Any] | None = None,
    ) -> None:
        """Initializes a new instance of the KernelFunctionFromMethod class.

        Args:
            method (Callable[..., Any]): The method to be called
            plugin_name (str | None): The name of the plugin
            stream_method (Callable[..., Any] | None): The stream method for the function
            parameters (list[KernelParameterMetadata] | None): The parameters of the function
            return_parameter (KernelParameterMetadata | None): The return parameter of the function
            additional_metadata (dict[str, Any] | None): Additional metadata for the function
        """
        if method is None:
            raise FunctionInitializationError("Method cannot be `None`")

        if not hasattr(method, "__kernel_function__") or method.__kernel_function__ is None:
            raise FunctionInitializationError("Method is not a Kernel function")

        # all these fields are created when the kernel function decorator is used,
        # so no need to check before using, will raise an exception if not set
        function_name = method.__kernel_function_name__  # type: ignore
        description = method.__kernel_function_description__  # type: ignore
        if parameters is None:
            parameters = [KernelParameterMetadata(**param) for param in method.__kernel_function_parameters__]  # type: ignore
        if return_parameter is None:
            return_parameter = KernelParameterMetadata(
                name="return",
                description=method.__kernel_function_return_description__,  # type: ignore
                default_value=None,
                type_=method.__kernel_function_return_type__,  # type: ignore
                type_object=method.__kernel_function_return_type_object__,  # type: ignore
                is_required=method.__kernel_function_return_required__,  # type: ignore
            )

        try:
            metadata = KernelFunctionMetadata(
                name=function_name,
                description=description,
                parameters=parameters,
                return_parameter=return_parameter,
                is_prompt=False,
                is_asynchronous=isasyncgenfunction(method) or iscoroutinefunction(method),
                plugin_name=plugin_name,
                additional_properties=additional_metadata if additional_metadata is not None else {},
            )
        except ValidationError as exc:
            # reraise the exception to clarify it comes from KernelFunction init
            raise FunctionInitializationError("Failed to create KernelFunctionMetadata") from exc

        args: dict[str, Any] = {
            "metadata": metadata,
            "method": method,
            "stream_method": (
                stream_method
                if stream_method is not None
                else method
                if isasyncgenfunction(method) or isgeneratorfunction(method)
                else None
            ),
        }

        super().__init__(**args)

    async def _invoke_internal(
        self,
        context: FunctionInvocationContext,
    ) -> None:
        """Invoke the function with the given arguments."""
        function_arguments = self.gather_function_parameters(context)
        result = self.method(**function_arguments)
        if isasyncgen(result):
            result = [x async for x in result]
        elif isawaitable(result):
            result = await result
        elif isgenerator(result):
            result = list(result)
        if not isinstance(result, FunctionResult):
            result = FunctionResult(
                function=self.metadata,
                value=result,
                metadata={"arguments": context.arguments, "used_arguments": function_arguments},
            )
        context.result = result

    async def _invoke_internal_stream(self, context: FunctionInvocationContext) -> None:
        if self.stream_method is None:
            raise NotImplementedError("Stream method not implemented")
        function_arguments = self.gather_function_parameters(context)
        context.result = FunctionResult(function=self.metadata, value=self.stream_method(**function_arguments))

    def _parse_parameter(self, value: Any, param_type: Any) -> Any:
        """Parses the value into the specified param_type, including handling lists of types."""
        if isinstance(param_type, type) and hasattr(param_type, "model_validate"):
            try:
                return param_type.model_validate(value)
            except Exception as exc:
                raise FunctionExecutionException(
                    f"Parameter is expected to be parsed to {param_type} but is not."
                ) from exc
        elif hasattr(param_type, "__origin__") and param_type.__origin__ is list:
            if isinstance(value, list):
                item_type = param_type.__args__[0]
                return [self._parse_parameter(item, item_type) for item in value]
            raise FunctionExecutionException(f"Expected a list for {param_type}, but got {type(value)}")
        else:
            try:
                if isinstance(value, dict) and hasattr(param_type, "__init__"):
                    return param_type(**value)
                return param_type(value)
            except Exception as exc:
                raise FunctionExecutionException(
                    f"Parameter is expected to be parsed to {param_type} but is not."
                ) from exc

    def gather_function_parameters(self, context: FunctionInvocationContext) -> dict[str, Any]:
        """Gathers the function parameters from the arguments."""
        function_arguments: dict[str, Any] = {}
        for param in self.parameters:
            if param.name is None:
                raise FunctionExecutionException("Parameter name cannot be None")
            if param.name == "kernel":
                function_arguments[param.name] = context.kernel
                continue
            if param.name == "service":
                function_arguments[param.name] = context.kernel.select_ai_service(self, context.arguments)[0]
                continue
            if param.name == "execution_settings":
                function_arguments[param.name] = context.kernel.select_ai_service(self, context.arguments)[1]
                continue
            if param.name == "arguments":
                function_arguments[param.name] = context.arguments
                continue
            if param.name in context.arguments:
                value: Any = context.arguments[param.name]
                if (
                    param.type_
                    and "," not in param.type_
                    and param.type_object
                    and param.type_object is not inspect._empty
                    and param.type_object is not Any
                ):
                    try:
                        value = self._parse_parameter(value, param.type_object)
                    except Exception as exc:
                        raise FunctionExecutionException(
                            f"Parameter {param.name} is expected to be parsed to {param.type_object} but is not."
                        ) from exc
                function_arguments[param.name] = value
                continue
            if param.is_required:
                raise FunctionExecutionException(
                    f"Parameter {param.name} is required but not provided in the arguments."
                )
            logger.debug(f"Parameter {param.name} is not provided, using default value {param.default_value}")
        return function_arguments

File: nodejs/semantic_kernel/functions/kernel-function-from-method.ts
import { FunctionExecutionException, FunctionInitializationError } from '../exceptions/function-exceptions'
import type { FunctionInvocationContext } from '../filters/functions/function-invocation-context'
import { FunctionResult } from './function-result'
import { KernelFunction } from './kernel-function'
import { KernelFunctionMetadata } from './kernel-function-metadata'
import { KernelParameterMetadata } from './kernel-parameter-metadata'

/**
 * Semantic Kernel Function from a method.
 */
export class KernelFunctionFromMethod extends KernelFunction {
  method: (...args: any[]) => any
  streamMethod?: ((...args: any[]) => any) | null = null

  constructor(params: {
    method: (...args: any[]) => any
    name?: string
    description?: string
    pluginName?: string | null
    streamMethod?: ((...args: any[]) => any) | null
    parameters?: KernelParameterMetadata[] | null
    returnParameter?: KernelParameterMetadata | null
    additionalMetadata?: Record<string, any> | null
  }) {
    const {
      method,
      name,
      description,
      pluginName = null,
      streamMethod = null,
      parameters = null,
      returnParameter = null,
      additionalMetadata = null,
    } = params

    if (!method) {
      throw new FunctionInitializationError('Method cannot be null')
    }

    // Check if method has kernel function metadata
    const methodAny = method as any
    if (!methodAny.__kernel_function__) {
      // If no kernel function metadata, use provided name and description
      // This allows creating functions from plain methods
      if (!name) {
        throw new FunctionInitializationError('Method is not a Kernel function and no name was provided')
      }
    }

    // Extract function metadata from decorated method or use provided values
    const functionName = name || methodAny.__kernel_function_name__
    const functionDescription = description || methodAny.__kernel_function_description__ || ''

    let functionParameters: KernelParameterMetadata[]
    if (parameters === null) {
      if (methodAny.__kernel_function_parameters__) {
        functionParameters = methodAny.__kernel_function_parameters__.map(
          (param: any) => new KernelParameterMetadata(param)
        )
      } else {
        functionParameters = []
      }
    } else {
      functionParameters = parameters
    }

    let functionReturnParameter: KernelParameterMetadata
    if (returnParameter === null) {
      if (methodAny.__kernel_function_return_description__) {
        functionReturnParameter = new KernelParameterMetadata({
          name: 'return',
          description: methodAny.__kernel_function_return_description__,
          defaultValue: null,
          type: methodAny.__kernel_function_return_type__,
          typeObject: methodAny.__kernel_function_return_type_object__,
          isRequired: methodAny.__kernel_function_return_required__,
        })
      } else {
        functionReturnParameter = new KernelParameterMetadata({
          name: 'return',
          description: '',
          isRequired: false,
        })
      }
    } else {
      functionReturnParameter = returnParameter
    }

    // Check if method is async
    const isAsynchronous =
      method.constructor.name === 'AsyncFunction' || method.constructor.name === 'AsyncGeneratorFunction'

    let metadata: KernelFunctionMetadata
    try {
      metadata = new KernelFunctionMetadata({
        name: functionName,
        description: functionDescription,
        parameters: functionParameters,
        returnParameter: functionReturnParameter,
        isPrompt: false,
        isAsynchronous,
        pluginName: pluginName ?? undefined,
        additionalProperties: additionalMetadata || {},
      })
    } catch (error) {
      console.error('Error creating KernelFunctionMetadata:', error)
      throw new FunctionInitializationError('Failed to create KernelFunctionMetadata')
    }

    super(metadata)

    this.method = method
    this.streamMethod =
      streamMethod !== null
        ? streamMethod
        : method.constructor.name === 'AsyncGeneratorFunction' || method.constructor.name === 'GeneratorFunction'
          ? method
          : null
  }

  /**
   * Invoke the function with the given arguments.
   */
  protected async _invokeInternal(context: FunctionInvocationContext): Promise<void> {
    const functionArguments = this.gatherFunctionParameters(context)
    let result = this.method(...Object.values(functionArguments))

    // Handle async generators
    if (result && typeof result[Symbol.asyncIterator] === 'function') {
      const items: any[] = []
      for await (const item of result) {
        items.push(item)
      }
      result = items
    }
    // Handle promises
    else if (result && typeof result.then === 'function') {
      result = await result
    }
    // Handle sync generators
    else if (result && typeof result[Symbol.iterator] === 'function') {
      result = Array.from(result)
    }

    if (!(result instanceof FunctionResult)) {
      result = new FunctionResult({
        function: this.metadata,
        value: result,
        metadata: { arguments: context.arguments, usedArguments: functionArguments },
      })
    }

    context.result = result
  }

  /**
   * Invoke the function with streaming.
   */
  protected async _invokeInternalStream(context: FunctionInvocationContext): Promise<void> {
    if (!this.streamMethod) {
      throw new Error('Stream method not implemented')
    }

    const functionArguments = this.gatherFunctionParameters(context)
    context.result = new FunctionResult({
      function: this.metadata,
      value: this.streamMethod(...Object.values(functionArguments)),
    })
  }

  /**
   * Parses the value into the specified param type, including handling lists of types.
   */
  private _parseParameter(value: any, paramType: any): any {
    // Handle Any or object type explicitly
    if (paramType === undefined || paramType === null || paramType === Object) {
      return value
    }

    // Handle Pydantic-like models with model_validate
    if (typeof paramType === 'function' && typeof (paramType as any).model_validate === 'function') {
      try {
        return (paramType as any).model_validate(value)
      } catch (error) {
        console.error('Error validating parameter with model_validate:', error)
        throw new FunctionExecutionException(`Parameter is expected to be parsed to ${paramType} but is not.`)
      }
    }

    // Handle Array types
    if (Array.isArray(value) && paramType === Array) {
      return value
    }

    // Try to construct the type
    try {
      if (typeof value === 'object' && value !== null && typeof paramType === 'function') {
        return new paramType(value)
      }
      return paramType(value)
    } catch (error) {
      console.error('Error parsing parameter:', error)
      throw new FunctionExecutionException(`Parameter is expected to be parsed to ${paramType} but is not.`)
    }
  }

  /**
   * Gathers the function parameters from the arguments.
   */
  gatherFunctionParameters(context: FunctionInvocationContext): Record<string, any> {
    const functionArguments: Record<string, any> = {}

    for (const param of this.parameters) {
      if (!param.name) {
        throw new FunctionExecutionException('Parameter name cannot be null')
      }

      // Handle special parameter names
      if (param.name === 'kernel') {
        functionArguments[param.name] = context.kernel
        continue
      }
      if (param.name === 'service') {
        // Get the service from kernel
        functionArguments[param.name] = context.kernel.getService()
        continue
      }
      if (param.name === 'execution_settings') {
        // Execution settings would typically come from context or arguments
        functionArguments[param.name] = null
        continue
      }
      if (param.name === 'arguments') {
        functionArguments[param.name] = context.arguments
        continue
      }

      // Check if argument is provided
      if (context.arguments.has(param.name)) {
        let value: any = context.arguments.get(param.name)

        if (param.type && !param.type.includes(',') && param.typeObject && param.typeObject !== Object) {
          try {
            value = this._parseParameter(value, param.typeObject)
          } catch (error) {
            console.error(`Error parsing parameter ${param.name}:`, error)
            throw new FunctionExecutionException(
              `Parameter ${param.name} is expected to be parsed to ${param.typeObject} but is not.`
            )
          }
        }

        functionArguments[param.name] = value
        continue
      }

      // Check if parameter is required
      if (param.isRequired) {
        throw new FunctionExecutionException(`Parameter ${param.name} is required but not provided in the arguments.`)
      }

      console.debug(`Parameter ${param.name} is not provided, using default value ${param.defaultValue}`)
    }

    return functionArguments
  }
}
